### ISSUE TITLE
RFC: Add `.raw` support to Builder instances analogous to `.whereRaw` et al.

### ISSUE THREAD

**Issue by tylerjwatson** [tgriesser/knex#3032]
# Environment

Knex version: any
Database + version: Azure SQL
OS: Linux

# Feature discussion / request

Hello Knex,

I would like to request that `.raw` be added to Builder instances for the purposes of appending dialect-specific statements to the end of built queries.

1. Explain what is your use case

Hold onto your hats Ladies and Gentlemen.

My Knex instance is responsible for connecting to, and managing an SQL Azure database full of temporal tables (I'm the guy who wrote #2857).  Temporal tables in MSSQL allow you to roll-back the database and view it at any point in time and you access historical data by appending a `FOR SYSTEM_TIME AS OF [date]` statement to the end of your queries.

If the user has indicated they want to time-travel the database, then the program should apply this temporal statement to all of the outgoing queries. 

Consider the following Express API pseudo:

```js
function temporalMiddleware(req, res, next) {
    req.db.on('start', query => {
      if (req.timeTravelEnabled && req.timeTravelDate) {
         query.raw('FOR SYSTEM_TIME AS OF ?', req.timeTravelDate);
      }
    });
}
```
 Now, I know the documentation discourages modifying a builder in event emitters, but _the use of the start hook really is the best approach for this scenario_.  `knex.raw` would be well suited to this scenario if it did not return a new builder instance.  You cannot modify the `query` reference passed to the event listener.

My database API involves a lot of dialect-specific builder functions and a lot of them pass around a Knex instance specifically for the purpose of using Knex.raw: 

```js
function applySpecifics(knex) {
  return query => knex.raw('? SOMETHING_SPECIFIC', [query]);
}
```

Avoiding this deletes a fair bit of code smell in userland APIs:

```js
function applySpecifics() {
  return query => query.raw('? SOMETHING_SPECIFIC');
}
```

OR:

A context object be provided to event listeners to allow the `builder` property to be replaced with a different builder reference so that `knex.raw` may be used.  This is the easiest option, but is a fairly horrid breaking change which I can't imaging being popular.

2. Explain what kind of feature would support this

Builder instances would support `.raw` methods which append raw statements to the built query when compiled.

3. Give some API proposal, how the feature should work

The function synonymous with `knex.raw` be placed on `Builder.prototype` which allows raw statements to be appended to built statements at compile time.  The API should support bindings in the traditional fashion.

I'm happy to implement the feature, however I'm a bit of a fledgling when it comes to the query compiler and builder architecture, so let me know what you think.

**Comment by elhigu** [comment]
tl;dr I would like to see something like this in knex, proposed API is a bit too ambiguous though and needs more discussion.

As far as I understood, the main point here was to have kind of `builder.raw` method, which will actually modify builder by wrapping query to a `.raw` in the last phase when building the query.

Some time ago there was almost the same  kind of feature, which was asking for a way to prepend stuff to the built query (to be able to pass some magic comments to a proxy server). This would have worked for that case too. 

Only case where this would not work, would be when people are trying to use this in a middle of a builder chain and thinking that those raw parts would be included in the middle of building a query, which is fine if this is documented well.

> A context object be provided to event listeners to allow the `builder` property to be replaced with a different builder reference so that `knex.raw` may be used. This is the easiest option, but is a fairly horrid breaking change which I can't imaging being popular.

I don't see how that would be a breaking change and actually passing context to even listeners has been planned already to be done. However reassigning query builder in that phase has not been in a list of future features.

Anyways I generally like the idea for being able to wrap query to a raw builder as a last phase of execution. That would be really powerful last resort for modifying query when knex API doesn't support some DB specific feature, without having to resort to full knex.raw which ignores for example post processing results step.

Only thing I don't like here is the method name and API. `builder.raw` is really ambiguous name, I would prefer something that states clearly that it is applied to the query builder after everything else is built. I cannot come up immediately with anything good myself though `.wrapToRaw`, `.finalRaw` doesn't sound really nice event that they are more explicit.

Second thing about API is that I don't like the fact that there is an implicit `?` binding to position original query for wrapping it to raw. Otherwise positioning parameters and query would be ambiguous... So I would prefer something explicit here too like:

```
query.finalRaw(
  'ANY THING ? IN THE WORLD WITH MORE BINDINGS ? ?', 
  [param1, query, param2], 
  {...options}
)
```

We should probably validate that one of the passed parameters is the same query.... or if not this would be even more powerful, since it could be used to rewrite old query builder with basically anything (so maybe an options object would be useful). Another thing where options object could be useful is to select if result from query should be treated as .raw query or as a builder query... normal knex.raw returns the result in a different way...

Then as a side note: if we would allow `.modify` to replace the whole query builder it would probably work for your use case too... that would be a separate feature request / discussion which can be opened if anyone cares to explore that possibility.

**Comment by wubzz** [comment]
I think adding functions that turn the builder into `Raw` doesn't make much sense, especially not if the `query` (builder) parameter is not used. Very confusing.

The example posted above is how I usually wrap builders if there's something I need that isn't already in the API:

```javascript
return query => knex.raw('? SOMETHING_SPECIFIC', [query]);
```

Sounds like your issue is not having access to `Raw` without passing around the `knex` object which could be solved by either extending the builder API with a getter for `Raw` or simply using `builder.client.raw`, although undocumented..

**Comment by tylerjwatson** [comment]
Perhaps consider `.prependRaw` and `.appendRaw` instead.

On Feb 5, 2019 4:46 PM, Mikael Lepistö <notifications@github.com> wrote:

tl;dr I would like to see something like this in knex, proposed API is a bit too ambiguous though and needs more discussion.

As far as I understood, the main point here was to have kind of builder.raw method, which will actually modify builder by wrapping query to a .raw in the last phase when building the query.

Some time ago there was almost the same kind of feature, which was asking for a way to prepend stuff to the built query (to be able to pass some magic comments to a proxy server). This would have worked for that case too.

Only case where this would not work, would be when people are trying to use this in a middle of a builder chain and thinking that those raw parts would be included in the middle of building a query, which is fine if this is documented well.

A context object be provided to event listeners to allow the builder property to be replaced with a different builder reference so that knex.raw may be used. This is the easiest option, but is a fairly horrid breaking change which I can't imaging being popular.

I don't see how that would be a breaking change and actually passing context to even listeners has been planned already to be done. However reassigning query builder in that phase has not been in a list of future features.

Anyways I generally like the idea for being able to wrap query to a raw builder as a last phase of execution. That would be really powerful last resort for modifying query when knex API doesn't support some DB specific feature, without having to resort to full knex.raw which ignores for example post processing results step.

Only thing I don't like here is the method name and API. builder.raw is really ambiguous name, I would prefer something that states clearly that it is applied to the query builder after everything else is built. I cannot come up immediately with anything good myself though .wrapToRaw, .finalRaw doesn't sound really nice event that they are more explicit.

Second thing about API is that I don't like the fact that there is an implicit ? binding to position original query for wrapping it to raw. Otherwise positioning parameters and query would be ambiguous... So I would prefer something explicit here too like:

query.finalRaw(
  'ANY THING ? IN THE WORLD WITH MORE BINDINGS ? ?',
  [param1, query, param2],
  {...options}
)


We should probably validate that one of the passed parameters is the same query.... or if not this would be even more powerful, since it could be used to rewrite old query builder with basically anything (so maybe an options object would be useful). Another thing where options object could be useful is to select if result from query should be treated as .raw query or as a builder query... normal knex.raw returns the result in a different way...

Then as a side note: if we would allow .modify to replace the whole query builder it would probably work for your use case too... that would be a separate feature request / discussion which can be opened if anyone cares to explore that possibility.

**Comment by elhigu** [comment]
This is a bit related to what`.wrap()` does. Wrap could be actually modified to support bindings and in that case it would do pretty much the same that this FR I suppose.

**Comment by willsoto** [comment]
@elhigu I am working on this now, but I am curious: is it possible to know where in the query builder chain a particular method was called?

I am looking to address the following point:

> ...use this in a middle of a builder chain and thinking that those raw parts would be included in the middle of building a query...

I think this API would be much more useful if the user had the ability to inject arbitrary raw expressions into the query as it is being built. 

Here is another - somewhat contrived - example:

```js
knex('books')
  .insert({title: 'Slaughterhouse Five'})
  .raw('on conflict (??) do update', ['title'])
  .raw('set updated_at = ?', [new Date()])
  .returning('*')
```

would output:

```sql
insert into ... on conflict ("title") do update set updated_at = 'date' returning *
```

In my implementation, raw expressions only get appended, but I don't know if there is a way to distinguish when it was called was called in the chain.

Any thoughts on this?

**Comment by elhigu** [comment]
Afaik the data is just accumulated to the query builder instance on every call. So there is no way to separate each `.raw` call after they are merged together and they will be appended to the query builder state in the same order how calls to `.raw` 
were executed.

**Comment by mercmobily** [comment]
This was a lot of work to work out.
I really want to leave this open as a feature request. However, it's a lot more involved than anybody would like. Here is a final writeup.

# RFC: Positional Raw Fragments on Query Builders (follow‑up to #3032 / PR #3593)

## Summary

Request: allow appending dialect‑specific raw SQL at well‑defined positions in the compiled query (e.g. after FROM, before WHERE), via a builder API, without dropping down to full raw SQL for the whole statement.

Primary motivator: MSSQL temporal tables need `FOR SYSTEM_TIME AS OF ?` immediately after a table reference. Users would like to attach that fragment consistently (e.g. in middleware) without rewriting queries.

This RFC documents the problem, current capabilities, lessons from PR #3593, and a practical, incremental proposal that fits Knex’s philosophy.

## Background

- Issue #3032 asked for a generic `builder.raw(...)` analogous to `.whereRaw`, to append raw text “at the end” of a query. The motivating example was MSSQL temporal tables.
- PR #3593 attempted an implementation by introducing an append‑only `raws` group.
- Maintainer feedback (see PR #3593 discussion):
  - A naïve “append at end” is insufficient; users need to be able to place raw fragments at specific stages (after FROM, before WHERE, etc.).
  - The ideal long‑term solution is an internal AST/graph for the query so insertion points are explicit and safe.
  - Short of an AST, a pragmatic approach is a set of position‑specific raw hooks or a single API that takes a `position` option.
  - There are tricky binding/escaping considerations when combining nested raw fragments; the API must leave responsibility for correctness with the caller.

## Current Capabilities (today)

Knex already exposes position‑specific raw methods for many stages:

- `fromRaw(sql, bindings)` → injects raw into FROM/table position
- `joinRaw(sql, bindings)` → JOIN position
- `whereRaw`, `havingRaw`, `orderByRaw`, `unionRaw`, etc.

Example (MSSQL temporal tables):

```js
// Exactly where the temporal clause belongs: after the table reference
knex
  .fromRaw('MyTable FOR SYSTEM_TIME AS OF ?', [asOfDate])
  .select('*')
  .where('id', 123);
```

If you need to apply this globally (e.g. in middleware), prefer composing with `modify` and position‑specific methods instead of a generic catch‑all:

```js
// Pseudocode: apply temporal clause for all table scans when toggled
req.db.on('start', (qb) => {
  if (!req.timeTravelEnabled || !req.timeTravelDate) return;
  qb.modify((b) => {
    // If the builder already has a concrete table, rebuild FROM as raw
    if (b._single && b._single.table) {
      const tbl = b._single.table;
      b._single.table = undefined;
      b.fromRaw('?? FOR SYSTEM_TIME AS OF ?', [tbl, req.timeTravelDate]);
    }
  });
});
```

This stays within supported APIs and avoids a free‑floating, end‑append `.raw()` that may land in the wrong place.

## Why a generic `.raw()` is problematic

- Position matters. Appending at the end breaks many syntaxes (e.g. MSSQL temporal clause must be after the table; placing it after WHERE/ORDER BY is wrong).
- Binding semantics. Composing nested raws easily leads to confusing placeholder/binding order; escaping `?` vs placeholders is already nuanced.
- Cross‑dialect confusion. A generic “append” suggests portability that often doesn’t exist.

## Proposal: Positional Raw API (incremental)

Introduce a narrowly scoped builder API that allows raw insertion at well‑defined stages. Two viable shapes:

1) A single method with a `position` option:

```js
knex('users')
  .select('*')
  .raw('FOR SYSTEM_TIME AS OF ?', [asOf], { position: 'afterFrom' })
  .where('id', 1);
```

2) Minimal new helpers (prefer small surface):

- `rawAfterFrom(sql, bindings)`
- `rawBeforeWhere(sql, bindings)`

Under the hood, these record stage‑scoped raw fragments and render them from the compiler at the appropriate moment, similar to how `whereRaw`/`havingRaw` are handled now.

Notes:
- This does not replace existing `…Raw` methods; it complements gaps like “after FROM”.
- The compiler must define a small, explicit set of insertion stages (e.g. beforeFrom, afterFrom, beforeWhere, afterWhere, beforeOrder).
- Dialects can choose to ignore unsupported positions (throw) to avoid pretending portability.
- Bindings are passed through as‑is; the caller owns correctness.

## Out of Scope

- A free‑floating “append at end” `.raw()` that tries to work for every dialect.
- A complete AST refactor (tracked separately in Schema V2 / query planner discussions).

## Alternatives

- Rely solely on existing `…Raw` methods and composition (`fromRaw`, `whereRaw`, etc.). This works today for many cases including temporal tables.
- Long‑term: adopt an internal AST for the query, enabling clean insertion anywhere. This is a larger effort outside the scope of this RFC; see prior maintainer notes.

## Risks & Mitigations

- Ambiguity about positions → Keep the enum small and well‑documented; throw on unsupported placements.
- Binding confusion → Document that bindings are not re‑ordered; raw fragments are rendered verbatim at the chosen stage.
- API sprawl → Prefer the single method with a `position` option over many new methods.

## Action Items

- Agree on an initial set of positions: `afterFrom`, `beforeWhere`, `afterWhere` (minimal to start).
- Implement stage‑scoped storage in the builder and render from the compiler alongside existing groups.
- Add tests per dialect for at least one practical use‑case (MSSQL temporal clause after FROM).
- Document with examples and caveats.

## Conclusion

A generic builder `.raw()` is not safe without placement semantics. We already cover most needs with existing position‑specific raw methods. For the remaining gaps (notably “after FROM”), a small, explicit positional raw API provides the power users need without over‑promising portability, and it aligns with Knex’s existing patterns.
